pkg/deployment/agency/state.go

```diff
@@ -115,3 +119,70 @@ func (d *StateExists) UnmarshalJSON(bytes []byte) error {
 	*d = bytes != nil
 	return nil
 }
+
+type CollectionIterator func(db, col string, planCollection *StatePlanCollection, shard string, plan ShardServers, current ShardServers)
+
+func (s State) IterateOverCollections(i CollectionIterator) {
+	for db, collections := range s.Plan.Collections {
+		for collection, details := range collections {
+			for shard, shardDetails := range details.Shards {
+				currShard := s.Current.Collections[db][collection][shard]
+
+				i(db, collection, &details, shard, shardDetails, currShard.Servers)
+			}
+		}
+	}
+}
+
+func (s State) IsDBServerInSync(serverID string) bool {
+	isInSync := true
+	s.IterateOverCollections(func(db, col string, planCollection *StatePlanCollection, shard string, plan ShardServers, current ShardServers) {
+		if !isInSync {
+			return
+		}
+		serverIsNotInSync := plan.Contains(serverID) && !current.Contains(serverID)
+		if serverIsNotInSync {
+			isInSync = false
+			return
+		}
+	})
+
+	return isInSync
+}
+
+func (s State) IsDBServerReadyToRestart(serverID string) bool {
+	readyToRestart := true
+
+	s.IterateOverCollections(func(db, col string, planCollection *StatePlanCollection, shard string, plan ShardServers, current ShardServers) {
+		if !readyToRestart {
+			return
+		}
+
+		wc := planCollection.GetWriteConcern(0)
+		rf := planCollection.GetReplicationFactor(0)
+
+		if wc >= len(plan) && len(plan) != 0 {
+			wc = len(plan) - 1
+		}
+
+		// Allow one server to be restarted if replication factor == write concern
+		if rf > 0 && wc == rf {
+			wc -= 1
+		}
+
+		serverIsNotInSync := plan.Contains(serverID) && !current.Contains(serverID)
```

Comment:
Move check !plan.contains at top - ignore shard if server is not desired for this shard

Reply:
done